### PR TITLE
operator: ProvisionToken support

### LINKED ISSUE #21417: operator: support ProvisionToken resources
What would you like Teleport to do?

The teleport-operator to support `ProvisionToken` resources.

What problem does this solve?

Administrators can create tokens in an Infrastructure-as-code fashion using the same tooling as the one they use to deploy Teleport. This improves feature parity between the teleport-operator and the Terraform provider.

If a workaround exists, please include it.

### PR DESCRIPTION
**Pull request by hugoShaka**
Fixes: https://github.com/gravitational/teleport/issues/21417

This PR adds a new `TeleportProvisionToken` CRD reconciled by the kubernetes operator.
This will be used when splitting auth and proxies ([see RFD 0096](https://github.com/gravitational/teleport/pull/18274)) to create kubernetes provision tokens ([see RFD 0094](https://github.com/gravitational/teleport/pull/17905)) for the proxies to join.

The CRD generator was not registering nested messages properly, the first commit contains the fix. (it was also swallowing errors)

Note for reviewers: the PR looks huge but you can disregard the big YAML manifests which are not really relevant. I'd love to find a way to improve readability of those PRs.

Blocked by https://github.com/gravitational/teleport/pull/18659

This PR should be rebased and CRD generation ran again with the new Kubernetes provision tokens.

### CODE REVIEW

File: operator/controllers/resources/provision_token_controller.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	saml, err := teleportClient.GetToken(ctx, name)
+	return saml, trace.Wrap(err)
+}
+
+// Create creates a Teleport saml_connector
+func (r provisionTokenClient) Create(ctx context.Context, saml types.ProvisionToken) error {
```
Comment (marcoandredinis) [inline operator/controllers/resources/provision_token_controller.go:47]: We should rename the argument to something like `token` right?

File: operator/controllers/resources/provision_token_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+			Name:      name,
+			Namespace: g.setup.namespace.Name,
+		},
+		Spec: resourcesv2.TeleportProvisionTokenSpec(*tokenSpec),
+	}
+	return trace.Wrap(g.setup.k8sClient.Create(ctx, token))
+}
+
+func (g *tokenTestingPrimitives) deleteKubernetesResource(ctx context.Context, name string) error {
+	saml := &resourcesv2.TeleportProvisionToken{
```
Comment (marcoandredinis) [inline operator/controllers/resources/provision_token_controller_test.go:100]: Same here.
I assume this was copy-pasted from the saml connector

File: operator/controllers/resources/provision_token_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	saml.Spec.Roles = []types.SystemRole{types.RoleNode, types.RoleProxy}
+	return trace.Wrap(g.setup.k8sClient.Update(ctx, saml))
+}
+
+func (g *tokenTestingPrimitives) compareTeleportAndKubernetesResource(tResource types.ProvisionToken, kubeResource *resourcesv2.TeleportProvisionToken) (bool, string) {
+	teleportMap, _ := teleportResourceToMap(tResource)
+	kubernetesMap, _ := teleportResourceToMap(kubeResource.ToTeleport())
+
+	// Signing key pair is populated server-side here
+	delete(teleportMap["spec"].(map[string]interface{}), "signing_key_pair")
```
Comment (strideynet) [inline operator/controllers/resources/provision_token_controller_test.go:133]: Is this leftover from the SAML implementation ?

**Review by strideynet — APPROVED** [review]
Happy with this once my points & those by Marco have been addressed.

File: api/types/provisioning.go [hunk 4]
```diff
@@ -69,6 +69,7 @@ func ValidateJoinMethod(method JoinMethod) error {
 // ProvisionToken is a provisioning token
 type ProvisionToken interface {
 	Resource
```
Comment (mdwn) [inline api/types/provisioning.go:71]: `ResourceWithOrigin` includes `Resource`, so I imagine that you can remove the `Resource` line here.

**Review by mdwn — APPROVED** [review]
One small comment, but otherwise pending resolution of the existing feedback, LGTM.

**Comment by hugoShaka** [comment]
I discovered that tokens, unlike most resources, are forcefully given an expiry.

This happens [in the tokenToItem translation](https://github.com/gravitational/teleport/blob/master/lib/services/local/provisioning.go#L70-L87):

```golang
func (s *ProvisioningService) tokenToItem(p types.ProvisionToken) (*backend.Item, error) {
	if err := p.CheckAndSetDefaults(); err != nil {
		return nil, trace.Wrap(err)
	}
	if p.Expiry().IsZero() || p.Expiry().Sub(s.Clock().Now().UTC()) < time.Second {
		p.SetExpiry(s.Clock().Now().UTC().Add(defaults.ProvisioningTokenTTL))
	}
```

The operator design relies on the fact resources it creates won't expire. If a token expires, it won't be reconciled immediately (because we don't watch teleport resources). This might lead to a non-working token for up to 10 hours.

The PR will have to be adapted to support this, the easiest workaround would be to set expiry to a very far date, like in 1000 years.

**Comment by hugoShaka** [comment]
I cherry-picked the changes and applied them on an up-to-date `master`. Thanks to https://github.com/gravitational/teleport/pull/21133, we can now create expiry-less tokens in v13, which unlocks the PR. It won't be backportable to v12, but at least v13 users will benefit from the feature.

**Review by r0mant — APPROVED** [review]

**Pull request merged**